lib/shopify_cli/environment.rb

```diff
@@ -97,32 +116,29 @@ def self.use_spin?(env_variables: ENV)
       )
     end
 
-    def self.infer_spin?(env_variables: ENV)
-      env_variable_truthy?(
-        Constants::EnvironmentVariables::INFER_SPIN,
-        env_variables: env_variables
-      )
-    end
-
     def self.spin_url(env_variables: ENV)
-      if infer_spin?(env_variables: env_variables)
-        # TODO: Remove version check and delete spin-legacy branch
-        #  once spin2 becomes the installed "spin" binary by default
-        spin_version = %x(spin version 2> /dev/null).strip
-        if spin_version.start_with?("spin-")
-          # spin2
-          raise ShopifyCLI:: Abort, "SPIN_INSTANCE must be specified" unless ENV.key?("SPIN_INSTANCE")
-          %x(spin show -o fqdn 2> /dev/null).strip
-        else
-          # spin-legacy
-          %x(spin info fqdn 2> /dev/null).strip
-        end
+      override = spin_url_override(env_variables: env_variables)
+      return override unless override.nil?
+
+      spin_response = if env_variables.key?(
+        Constants::EnvironmentVariables::SPIN_INSTANCE
+      )
+        spin_show
       else
-        spin_workspace = spin_workspace(env_variables: env_variables)
-        spin_namespace = spin_namespace(env_variables: env_variables)
-        spin_host = spin_host(env_variables: env_variables)
-        "#{spin_workspace}.#{spin_namespace}.#{spin_host}"
+        spin_show(latest: true)
       end
+
+      begin
+        instance = JSON.parse(spin_response)
+        raise "Missing key 'fqdn' from spin show. Actual response: #{instance}" unless instance.include?("fqdn")
+        instance["fqdn"]
+      rescue => e
+        raise "Failed to infer spin environment from spin show response #{spin_response}: #{e}"
```

Comment:
We shouldn't run into these error cases, but these error messages should help identify that the spin tool is the culprit in the case that it starts to return bad output.